src/typegate/src/runtimes/patterns/worker_manager/mod.ts

```diff
@@ -120,14 +190,41 @@ export class BaseWorkerManager<
         return false;
       }
 
-      task.worker.destroy();
       this.#activeTasks.delete(taskId);
       this.#tasksByName.get(name)!.delete(taskId);
       // startedAt records are not deleted
 
+      if (destroy) {
+        task.worker.destroy();
+
+        const taskAdded = this.#waitQueue.shift(() => this.#workerFactory());
+        if (!taskAdded) { // no task from the queue
+          if (ensureMinWorkers) {
+            const { minWorkers } = this.#poolConfig;
+            if (minWorkers != null && this.#workerCount < minWorkers) {
+              this.#idleWorkers.push(this.#workerFactory());
+            }
+          }
+        }
+      } else {
+        const taskAdded = this.#waitQueue.shift(() => task.worker);
+        if (!taskAdded) { // worker has not been reassigned
+          const { maxWorkers } = this.#poolConfig;
+          // how?? xD
+          // We might add "urgent" tasks in the future;
+          // in this case the worker count might exceed `maxWorkers`.
+          if (maxWorkers != null && this.#workerCount >= maxWorkers) {
+            task.worker.destroy();
+          } else {
+            this.#idleWorkers.push(task.worker);
```

Comment:
I feel like we should never exceed max but instead ensure it is big enough at start. Not sure how that would fit here but a design idea is to allocate as much as needed with a small room for newer tasks: allocate `minAlloc`, then if that is not enough and then task is urgent then allocate` minAlloc`+5, (`minAlloc`+5)+5, ..., min(`minAlloc` + 5 + 5 + ...,  `maxAlloc`) etc. 5 is completely arbitrary btw.